Unable to upload big files to S3 storage

When uploading big (128M) file to Artipie backed with S3 (OBS) storage, it returns 500 error with S3 SDK stacktrace:
```
java.util.concurrent.CompletionException: software.amazon.awssdk.services.s3.model.S3Exception: The specified method is not allowed against this resource. (Service: S3, Status Code: 405, Request ID: 000001727ECBCB598013AC417BF795C9)
java.util.concurrent.CompletionException: software.amazon.awssdk.services.s3.model.S3Exception: The specified method is not allowed against this resource. (Service: S3, Status Code: 405, Request ID: 000001727ECBCB598013AC417BF795C9)
	at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException(CompletableFutureUtils.java:61)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncExecutionFailureExceptionReportingStage.lambda$execute$0(AsyncExecutionFailureExceptionReportingStage.java:51)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
	at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:75)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.retryResponseIfNeeded(AsyncRetryableStage.java:156)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.retryIfNeeded(AsyncRetryableStage.java:120)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.lambda$execute$0(AsyncRetryableStage.java:107)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$executeHttpRequest$1(MakeAsyncHttpRequestStage.java:138)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: software.amazon.awssdk.services.s3.model.S3Exception: The specified method is not allowed against this resource. (Service: S3, Status Code: 405, Request ID: 000001727ECBCB598013AC417BF795C9)
	at software.amazon.awssdk.services.s3.model.S3Exception$BuilderImpl.build(S3Exception.java:95)
	at software.amazon.awssdk.services.s3.model.S3Exception$BuilderImpl.build(S3Exception.java:55)
	at software.amazon.awssdk.protocols.query.unmarshall.AwsXmlErrorProtocolUnmarshaller.handle(AwsXmlErrorProtocolUnmarshaller.java:127)
	at software.amazon.awssdk.protocols.query.unmarshall.AwsXmlErrorProtocolUnmarshaller.handle(AwsXmlErrorProtocolUnmarshaller.java:86)
	at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0(AsyncResponseHandler.java:88)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)
	at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete(AsyncResponseHandler.java:129)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:180)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.access$500(ResponseHandler.java:67)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onComplete(ResponseHandler.java:296)
	at com.typesafe.netty.HandlerPublisher.complete(HandlerPublisher.java:408)
	at com.typesafe.netty.HandlerPublisher.handlerRemoved(HandlerPublisher.java:395)
	at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:979)
	at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:477)
	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:423)
	at com.typesafe.netty.http.HttpStreamsHandler.removeHandlerIfActive(HttpStreamsHandler.java:328)
	at com.typesafe.netty.http.HttpStreamsHandler.handleReadHttpContent(HttpStreamsHandler.java:189)
	at com.typesafe.netty.http.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:165)
	at com.typesafe.netty.http.HttpStreamsClientHandler.channelRead(HttpStreamsClientHandler.java:148)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1486)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1235)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1282)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	... 1 more
```

--- a/src/main/java/com/artipie/asto/s3/S3Storage.java
+++ b/src/main/java/com/artipie/asto/s3/S3Storage.java
@@ -64,7 +64,13 @@ import software.amazon.awssdk.services.s3.model.S3Object;
  *  Also whole operation does not complete until abort() is complete.
  *  It would be better to finish save() operation right away and do abort() in background,
  *  but it makes testing the method difficult.
+ * @todo #193:30min Too many methods in `S3Storage` class.
+ *  `S3Storage` has too many methods.
+ *  It could be refactored to move some of it's logic to other classes.
+ *  In particular there three methods related to saving bytes
+ *  (multipart or not, selecting proper method). This could be moved out of this class.
  */
+@SuppressWarnings("PMD.TooManyMethods")
 public final class S3Storage implements Storage {
 
     /**
@@ -83,14 +89,33 @@ public final class S3Storage implements Storage {
     private final String bucket;
 
     /**
+     * Multipart allowed flag.
+     */
+    private final boolean multipart;
+
+    /**
      * Ctor.
      *
      * @param client S3 client.
      * @param bucket Bucket name.
      */
     public S3Storage(final S3AsyncClient client, final String bucket) {
+        this(client, bucket, true);
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param client S3 client.
+     * @param bucket Bucket name.
+     * @param multipart Multipart allowed flag.
+     *  <code>true</code> - if multipart feature is allowed for larger blobs,
+     *  <code>false</code> otherwise.
+     */
+    public S3Storage(final S3AsyncClient client, final String bucket, final boolean multipart) {
         this.client = client;
         this.bucket = bucket;
+        this.multipart = multipart;
     }
 
     @Override
@@ -134,44 +159,13 @@ public final class S3Storage implements Storage {
 
     @Override
     public CompletableFuture<Void> save(final Key key, final Content content) {
-        final Flowable<ByteBuffer> cache = Flowable.fromPublisher(content).cache();
-        return content.size()
-            .map(size -> CompletableFuture.completedFuture(Optional.of(size)))
-            .orElseGet(
-                () -> cache
-                    .map(Buffer::remaining)
-                    .scanWith(() -> 0L, (sum, item) -> sum + item)
-                    .takeUntil(total -> total >= S3Storage.MIN_MULTIPART)
-                    .lastOrError()
-                    .to(SingleInterop.get())
-                    .toCompletableFuture()
-                    .<Optional<Long>>thenApply(
-                        last -> {
-                            final Optional<Long> size;
-                            if (last >= S3Storage.MIN_MULTIPART) {
-                                size = Optional.empty();
-                            } else {
-                                size = Optional.of(last);
-                            }
-                            return size;
-                        }
-                    )
-            ).thenApply(
-                sizeOpt -> sizeOpt
-                    .<Content>map(size -> new Content.From(size, cache))
-                    .orElse(new Content.From(cache))
-            ).thenCompose(
-                updated -> {
-                    final CompletableFuture<Void> future;
-                    final Optional<Long> size = updated.size();
-                    if (size.isPresent() && size.get() < S3Storage.MIN_MULTIPART) {
-                        future = this.put(key, updated);
-                    } else {
-                        future = this.uploadMultipart(key, updated);
-                    }
-                    return future;
-                }
-            );
+        final CompletableFuture<Void> result;
+        if (this.multipart) {
+            result = this.saveWithMultipartAllowed(key, content);
+        } else {
+            result = this.put(key, content);
+        }
+        return result;
     }
 
     @Override
@@ -245,6 +239,55 @@ public final class S3Storage implements Storage {
     @Override
     public CompletableFuture<Transaction> transaction(final List<Key> keys) {
         throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Saves the bytes to the specified key.
+     * Uses multipart upload feature if size is greater or equal MIN_MULTIPART.
+     *
+     * @param key The key
+     * @param content Bytes to save
+     * @return Completion or error signal.
+     */
+    private CompletableFuture<Void> saveWithMultipartAllowed(final Key key, final Content content) {
+        final Flowable<ByteBuffer> cache = Flowable.fromPublisher(content).cache();
+        return content.size()
+            .map(size -> CompletableFuture.completedFuture(Optional.of(size)))
+            .orElseGet(
+                () -> cache
+                    .map(Buffer::remaining)
+                    .scanWith(() -> 0L, (sum, item) -> sum + item)
+                    .takeUntil(total -> total >= S3Storage.MIN_MULTIPART)
+                    .lastOrError()
+                    .to(SingleInterop.get())
+                    .toCompletableFuture()
+                    .<Optional<Long>>thenApply(
+                        last -> {
+                            final Optional<Long> size;
+                            if (last >= S3Storage.MIN_MULTIPART) {
+                                size = Optional.empty();
+                            } else {
+                                size = Optional.of(last);
+                            }
+                            return size;
+                        }
+                    )
+            ).thenApply(
+                sizeOpt -> sizeOpt
+                    .<Content>map(size -> new Content.From(size, cache))
+                    .orElse(new Content.From(cache))
+            ).thenCompose(
+                updated -> {
+                    final CompletableFuture<Void> future;
+                    final Optional<Long> size = updated.size();
+                    if (size.isPresent() && size.get() < S3Storage.MIN_MULTIPART) {
+                        future = this.put(key, updated);
+                    } else {
+                        future = this.uploadMultipart(key, updated);
+                    }
+                    return future;
+                }
+            );
     }
 
     /**

